src/librustc_typeck/diagnostics.rs

```diff
@@ -2804,6 +2814,15 @@ trait Trait {
 }
 
 type Foo = Trait<Bar=i32>; // ok!
+
+// or:
+
+trait Trait {
+    type Bar;
+    type Baz; // Baz has been defined before being used
```

Comment:
Hum... This sentence implies that a type must be defined before being used, which is wrong:

```rust
trait Foo {
    fn foo(&self, Self::Bar);

    type Bar;
}
```

What about:

```text
trait Trait {
    type Bar;
    type Baz; // we need the `Baz` type here.

    // and now we can use it here:
    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
```
?